### PR TITLE
 [ANE-2382] Add Pip constraint

### PR DESCRIPTION
**Pull request by james-fossa**
# Overview

Implement the Pip [Constraint](https://github.com/fossas/locator-rs/pull/14).

## Acceptance criteria

This should support [requirement specifiers](https://pip.pypa.io/en/stable/reference/requirement-specifiers/#requirement-specifiers) as much as possible. The maybe-notable exception is that the `===` operator is treated as an `Equal` constraint, instead of as a direct string comparison.

That's for two reasons.
- First, it doesn't make any sense for a security advisory - if `pip` resolves `0!1.2.3` identically to `1.2.3` then we're unlikely to gain anything by acting as though `0!1.2.3` is the only version with a vulnerability.
- Second, it would imply an extra kind of `Constraint` for a very niche case.

## Testing plan

Unit tests, and suggestions welcome: the more the merrier.

## Metrics

N/A

## Risks

There's a lot of potential edge-cases in [version specifiers](https://packaging.python.org/en/latest/specifications/version-specifiers/#version-specifiers), I might've missed some.

## References

- [Version Specifiers](https://packaging.python.org/en/latest/specifications/version-specifiers/#version-specifiers)
- [Requirement Specifiers](https://pip.pypa.io/en/stable/reference/requirement-specifiers/#requirement-specifiers)
- [ANE-2382](https://fossa.atlassian.net/browse/ANE-2382)

## Checklist

- [x] I added tests for this PR's change (or explained in the PR description why tests don't make sense).

[ANE-2382]: https://fossa.atlassian.net/browse/ANE-2382?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/constraint/pip.rs [hunk 1]
```diff
@@ -0,0 +1,735 @@
...
+                    ],
+                    pre_release,
+                    post_release: None,
+                    dev_release: None,
+                };
+
+                let pre_opt = &semver.pre;
+                if !pre_opt.is_empty() {
+                    let pre_str = pre_opt;
+                    let (kind, num) = match pre_str.as_str() {
```
Comment (james-fossa) [inline src/constraint/pip.rs:402]: This isn't exhaustive / I ought to go back and use `PreRelease::parse`

File: src/constraint/pip.rs [hunk 2]
```diff
@@ -0,0 +1,670 @@
...
+pub enum PipConstraintError {
+    #[display("ConstraintParseError({constraints}, {message})")]
+    ConstraintParseError {
+        constraints: String,
+        message: String,
+    },
+
+    #[display("VersionParseError({version}, {message})")]
+    VersionParseError { version: String, message: String },
+}
```
Comment (jssblck) [inline src/constraint/pip.rs:140]: [nit] you already know these are errors
[nit] error messages ideally read more like english
[nit] you're deriving `Error` with `thiserror` but using `derive_more::Display`, prefer to pick one or the other (`thiserror` or `derive_more`). the suggestion below assumes `thiserror`.
```suggestion
pub enum PipConstraintError {
    #[error("parse constraints {constraints:?}: {message:?})")]
    ConstraintParse {
        constraints: String,
        message: String,
    },

    #[error("parse version {version:?}: {message:?})")]
    VersionParse { version: String, message: String },
}
```

File: src/constraint/pip.rs [hunk 3]
```diff
@@ -0,0 +1,670 @@
...
+
+#[tracing::instrument]
+pub fn compare(
+    constraint: &Constraint,
+    fetcher: Fetcher,
+    target: &Revision,
+) -> Result<bool, PipConstraintError> {
+    if let (Revision::Semver(_), Revision::Semver(_)) = (constraint.revision(), target) {
+        Ok(fallback::compare(constraint, fetcher, target))
+    } else {
```
Comment (jssblck) [inline src/constraint/pip.rs:150]: [nit] prefer an early return here to reduce nesting
```suggestion
    if let (Revision::Semver(_), Revision::Semver(_)) = (constraint.revision(), target) {
        return Ok(fallback::compare(constraint, fetcher, target));
    }

```

File: src/constraint/pip.rs [hunk 4]
```diff
@@ -0,0 +1,670 @@
...
+#[cfg(test)]
+mod tests {
+    use simple_test_case::test_case;
+
+    use super::*;
+    use crate::{Revision, constraint};
+
+    const FETCHER: Fetcher = Fetcher::Pip;
+
+    #[test_case("1!2.3.4.a5", PipVersion { epoch: 1, release_segments: vec![2, 3, 4], pre_release: Some(PreRelease::Alpha(5)), post_release: None, dev_release: None }; "epoch1_alpha_short")]
```
Comment (jssblck) [inline src/constraint/pip.rs:582]: [nit] having `PipVersion` derive `bon::Builder` would make these tests a little cleaner to read- e.g. this line would look like this:
```rust
#[test_case("1!2.3.4.a5", PipVersion::builder().epoch(1).release_segments([2, 3, 4]).pre_release(PreRelease::Alpha(5)).build(); "epoch1_alpha_short")]
```

But it's fine to leave as is if desired.

File: src/constraint/pip.rs [hunk 5]
```diff
@@ -0,0 +1,670 @@
...
+    #[test_case("1.2.3", PipVersion { epoch: 0, release_segments: vec![1, 2, 3], pre_release: None, post_release: None, dev_release: None }; "simple_version")]
+    #[test_case("1!2.3.4_rc5", PipVersion { epoch: 1, release_segments: vec![2, 3, 4], pre_release: Some(PreRelease::RC(5)), post_release: None, dev_release: None }; "epoch1_prerelease")]
+    #[test_case("1.2.3_pre4", PipVersion { epoch: 0, release_segments: vec![1, 2, 3], pre_release: Some(PreRelease::RC(4)), post_release: None, dev_release: None }; "prerelease")]
+    #[test_case("1.2.3_a", PipVersion { epoch: 0, release_segments: vec![1, 2, 3], pre_release: Some(PreRelease::Alpha(0)), post_release: None, dev_release: None }; "implicit_prerelease")]
+    #[test_case("1.2.3-1", PipVersion { epoch: 0, release_segments: vec![1, 2, 3], pre_release: None, post_release: Some(1), dev_release: None }; "implicit_postrelease")]
+    #[test]
+    fn test_pip_version_parsing(input: &str, expected: PipVersion) {
+        let actual = PipVersion::parse(input).expect("should parse version");
+        dbg!(&expected, &actual);
```
Comment (jssblck) [inline src/constraint/pip.rs:597]: needed?

**Review by jssblck — APPROVED** [review]
Nice work!

File: src/constraint/pip.rs (hunk 4 above)
Comment (james-fossa) [reply src/constraint/pip.rs]: You're right that's much more legible

File: src/constraint/pip.rs (hunk 5 above)
Comment (james-fossa) [reply src/constraint/pip.rs:600]: ope!
